ROS2 migrate diagnostic aggregator

norro opened the pull request:
PR to integrate migration of diagnostic aggregator into migration of diagnostic updater. Ongoing, not yet ready to merge.

norro committed:
Removed AMENT_IGNORE and uncrustified

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Started migration of c++ API

* To be done: logging, assertions, parameter handling

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Started migration of python tests

Karsten1987 commented on diagnostic_aggregator/CMakeLists.txt:
```diff
@@ -29,46 +25,61 @@ add_library(${PROJECT_NAME}
+  "diagnostic_msgs"
+  "rclcpp"
+  "std_msgs"
+  "pluginlib"
```
alpha order

norro replied:
will adapt it accordingly

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/aggregator.hpp:
```diff
@@ -36,29 +36,31 @@
 #include <string>
 #include <map>
 #include <vector>
 #include <set>
```
alpha order

norro committed:
Started migration of analyzer group

* Migrated from XMLRPC to ROS2 parameters parsing
* Doesn't create working analzers, yet

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Migrated analyzers plugin

* Split anaylzers into seperate plugin lib
* Build shared lib to be used by plugin class loader
* Fixed plugin registration of analyzers

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Analyzer group correctly setting up analyzers

* Improved parameter handling of generic_analyzer

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
API migration to ROS2 c++ + logging

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
uncrustified

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro commented:
@Karsten1987 new status: not only compiling, but now setting up analyzers and analyzer groups according to config file, e.g.:
`ros2 run diagnostic_aggregator aggregator_node __params:=demo/pr2_analyzers.yaml`

I suggest merging this pull request now, and then creating a pull request for `ros/diagnostics` which we can use for (more public) discussion.
/cc @ralph-lange

norro commented:
GenericAnalyzer seems to be working as aspected, according to first tests.

norro committed:
Reworked analyzer paths and names

* Separate handling of the analyzer's (and analyzer group's):
** "nice" name
** path (path of their results in the robot monitor)
** breadcrumb (prefix of their yaml configuration)
* Logging
* Uncrustify

norro committed:
Examples

norro committed:
Less strict cpplint

norro committed:
removed using namespace

norro commented:
@Karsten1987 now doing what it is supposed to do (IMHO, example included). Please merge and then create a pull request to ROS 2.

Karsten1987 commented on diagnostic_aggregator/CMakeLists.txt:
```diff
@@ -1,74 +1,121 @@
+find_package(ament_cmake REQUIRED)
+find_package(ament_cmake_python REQUIRED)
+find_package(diagnostic_msgs REQUIRED)
+find_package(rclcpp REQUIRED)
```
nit: alphabetical order

Karsten1987 commented on diagnostic_aggregator/CMakeLists.txt:
```diff
@@ -1,74 +1,121 @@
+  $<INSTALL_INTERFACE:include>)
+ament_target_dependencies(${PROJECT_NAME}
+  "diagnostic_msgs"
+  "rclcpp"
```
nit: alphabetical order

Karsten1987 commented on diagnostic_aggregator/CMakeLists.txt:
```diff
@@ -1,74 +1,121 @@
+  "bondcpp"
+  "std_msgs"
+  "pluginlib"
+  "boost"
```
are you sure that boost is still being used? If so, you'd need a `find_package(Boost COMPONENTS ...)` before using it.

norro replied:
It was still in use (mutex, regex), however, I could swap it to stdlib and get rid of all boost dependencies.

Karsten1987 commented on diagnostic_aggregator/CMakeLists.txt:
```diff
@@ -1,74 +1,121 @@
+  DESTINATION include
 )
+
+target_compile_definitions(${PLUGIN_LIB} PRIVATE "DIAGNOSTICS_DEFAULT_ANALYZER_PLUGINS_BUILDING_LIBRARY")
```
consider shifting this call next to the `add_library` call for the target

Karsten1987 commented on diagnostic_aggregator/CMakeLists.txt:
```diff
@@ -1,74 +1,121 @@
+ament_export_dependencies(ament_cmake)
+ament_export_dependencies(ament_cmake_python)
+ament_export_dependencies(std_msgs)
+ament_export_dependencies(diagnostic_msgs)
```
alphabetical order

Karsten1987 commented on diagnostic_aggregator/demo/diag-pub.yaml:
```diff
@@ -0,0 +1,23 @@
+    discard:
+      type: diagnostic_aggregator/DiscardAnalyzer
+      path: DoesItNeedAPart?
+      regex: [ 'name\d' ]
```
there is no correct new line

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/aggregator.hpp:
```diff
@@ -36,29 +36,31 @@
 #include <set>
-#include <boost/shared_ptr.hpp>
+#include <memory>
 #include <boost/thread/mutex.hpp>
```
can that be replaced with a `std::mutex`?

norro replied:
Yes, we can! ;)

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/aggregator.hpp:
```diff
@@ -120,41 +122,51 @@ class Aggregator
-  double getPubRate() const { return pub_rate_; }
+  double getPubRate() const {return pub_rate_;}
+
+  rclcpp::Node::SharedPtr get_node();
```
can this function be const?

norro replied:
not sure, will try that

norro replied:
Since aggregator will become a node now (instead of just owning one), I think about only exposing its parameter interface via API. What do you think?

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/aggregator.hpp:
```diff
@@ -120,41 +122,51 @@ class Aggregator
+    std::shared_ptr<diagnostic_msgs::srv::AddDiagnostics::Response>);
 
-  AnalyzerGroup* analyzer_group_;
+  AnalyzerGroup * analyzer_group_;
```
does it make sense to use `std::unique_ptr` for these? Same below

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/aggregator.hpp:
```diff
@@ -175,29 +188,30 @@ class Aggregator
    *  which was created in the addDiagnostics function
    */
-  void bondFormed(boost::shared_ptr<Analyzer> group);
+  void bondFormed(std::shared_ptr<Analyzer>);
```
why removing the argument variable name?

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/analyzer.hpp:
```diff
@@ -33,30 +33,35 @@
-#include <diagnostic_msgs/DiagnosticStatus.h>
-#include <diagnostic_msgs/KeyValue.h>
-#include "diagnostic_aggregator/status_item.h"
+#include <string>
```
alpha order

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/analyzer.hpp:
```diff
@@ -88,60 +93,66 @@ class Analyzer
    */
-  virtual bool init(const std::string base_path, const ros::NodeHandle &n) = 0;
+  virtual bool init(
+    const std::string &,
```
why removing the argument variable names?

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/analyzer.hpp:
```diff
@@ -88,60 +93,66 @@ class Analyzer
+  virtual bool init(
+    const std::string &,
+    const std::string &,
+    const rclcpp::Node::SharedPtr) = 0;
```
I think it makes more sense to pass in the node ptr in the Analyzer constructor.

norro replied:
I think this is passed in `init()` instead of the constructor, as the constructor can't have any arguments to to the pluginlib (the `analyzer_loader` will raise errors, when constructor arguments are introduced).

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/analyzer.hpp:
```diff
@@ -88,60 +93,66 @@ class Analyzer
    */
   virtual std::string getName() const = 0;
 
+  rclcpp::Clock::SharedPtr clock_;
```
shouldn't this instance be private?

norro replied:
protected, yes, good point.

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/analyzer_group.hpp:
```diff
@@ -144,19 +146,19 @@ class AnalyzerGroup : public Analyzer
 
 private:
-  std::string path_, nice_name_;
+  std::string path_, nice_name_, breadcrumb_;
```
consider one variable per line

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/generic_analyzer.hpp:
```diff
@@ -33,70 +33,67 @@
-      std::string find = param[i];
-      output.push_back(find);
+  } else if (type == rclcpp::ParameterType::PARAMETER_STRING_ARRAY) {
+    auto param_array = param.as_string_array();
```
can we not directly assign output here?
```
output = param.as_string_array();
```

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/generic_analyzer_base.hpp:
```diff
@@ -65,195 +70,221 @@ namespace diagnostic_aggregator {
 
-    boost::shared_ptr<diagnostic_msgs::DiagnosticStatus> header_status(new diagnostic_msgs::DiagnosticStatus());
+    std::shared_ptr<diagnostic_msgs::msg::DiagnosticStatus> header_status(
+      new diagnostic_msgs::msg::DiagnosticStatus());
```
`auto header_status = std::make_shared<diagnostics_msgs::msg::DiagnosticStatus>()`?

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/generic_analyzer_base.hpp:
```diff
@@ -65,195 +70,221 @@ namespace diagnostic_aggregator {
+    while (it != items_.end()) {
       std::string name = it->first;
-      boost::shared_ptr<StatusItem> item = it->second;
+      std::shared_ptr<StatusItem> item = it->second;
```
consider `auto`

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/generic_analyzer_base.hpp:
```diff
@@ -65,195 +70,221 @@ namespace diagnostic_aggregator {
-    {
+    std::map<std::string, std::shared_ptr<StatusItem>>::iterator it = items_.begin();
+    while (it != items_.end()) {
       std::string name = it->first;
```
```suggestion
      std::string & name = it->first;
```

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/generic_analyzer_base.hpp:
```diff
@@ -65,195 +70,221 @@ namespace diagnostic_aggregator {
     // If we expect a given number of items, check that we have this number
-    if (num_items_expected_ == 0 && items_.size() == 0)
-    {
+    if (num_items_expected_ == 0 && items_.size() == 0) {
```
```suggestion
    if (num_items_expected_ == 0 && items_.empty()) {
```

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/generic_analyzer_base.hpp:
```diff
@@ -65,195 +70,221 @@ namespace diagnostic_aggregator {
-      header_status->level = std::max(header_status->level, level);
-      
-      diagnostic_msgs::KeyValue kv;
+      header_status->level = std::max(int8_t(header_status->level), level);
```
```suggestion
      header_status->level = std::max(static_cast<int8_t>(header_status->level), level);
```

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/generic_analyzer_base.hpp:
```diff
@@ -65,195 +70,221 @@ namespace diagnostic_aggregator {
       item << items_.size();
 
-      if (items_.size() > 0)
+      if (items_.size() > 0) {
```
```suggestion
      if (!items_.empty()) {
```

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/other_analyzer.hpp:
```diff
@@ -65,28 +70,39 @@ class OtherAnalyzer : public GenericAnalyzerBase
+  }
 
-  ~OtherAnalyzer() { }
+  ~OtherAnalyzer()
```
```suggestion
  virtual ~OtherAnalyzer()
```

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/other_analyzer.hpp:
```diff
@@ -95,44 +111,40 @@ class OtherAnalyzer : public GenericAnalyzerBase
+      // "Other" items are considered an error.
+      std::vector<std::shared_ptr<diagnostic_msgs::msg::DiagnosticStatus>>::iterator it =
+        processed.begin();
+      for (; it != processed.end(); ++it) {
```
consider `std::for_each` or simply `for (auto & msg : processed) {`

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/status_item.hpp:
```diff
@@ -122,36 +139,40 @@ inline std::string valToMsg(const int val)
   std::string output_name = input_name;
 
-  if (chaff.size() == 0)
+  if (chaff.size() == 0) {
```
```suggestion
  if (chaff.empty()) {
```

Karsten1987 commented on diagnostic_aggregator/include/diagnostic_aggregator/status_item.hpp:
```diff
@@ -161,71 +182,75 @@ class StatusItem
+    const std::string item_name, const std::string message = "Missing",
+    const DiagnosticLevel level = Level_Stale);
 
   ~StatusItem();
```
should be virtual?

Karsten1987 commented on diagnostic_aggregator/package.xml:
```diff
@@ -10,28 +10,26 @@
+  <build_depend>pluginlib</build_depend>
+
+  <depend>rclpy</depend>
+  <depend>bondpy</depend>
```
alphaorder

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -33,214 +33,285 @@
+ * @todo(anordman): make aggregator a lifecycle node.
+ */
+Aggregator::Aggregator()
+: n_(std::make_shared<rclcpp::Node>("analyzers",
```
I believe this should be templated on Node. Ideally the `Aggregator` should be able to take a `rclcpp::Node` as well as a `rclcpp_lifecycle::Node`.

norro replied:
I think we should rather go with `Aggregator` actually extending `rclcpp_lifecycle::Node`.

norro replied:
This is a bit more tricky than I initally thought:

1. It can't be `rclcpp_lifecycle::LifecycleNode` easily, because the API of the `Bond` formed between the aggregator and the analyzers expects a `rclcpp:Node`. This might require refactoring of `Bond`.
2. The aggregator probably _has_ a node (as opposed to _being_ a node), because said `Bonds` get `shared_ptr` to the aggregator and analyzers. If the aggregator would pass itself as shared_ptr, this would introduce a cyclic dependency since the aggregator node then would have `shared_ptrs` of the `Bonds`, and the `Bonds` would have `shared_ptrs` to the aggregator. This could maybe solved with the one of these pointers being `weak_ptrs`, but I'm not sure.

@Karsten1987 any idea?

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -33,214 +33,285 @@
+    "Retrieved %d parameter(s).",
+    parameters.size());
+  
+  for (auto & param : parameters) {
```
```suggestion
  for (const auto & param : parameters) {
```

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -33,214 +33,285 @@
-  if (!analyzer_group_->init(base_path_, nh))
-  {
-    ROS_ERROR("Analyzer group for diagnostic aggregator failed to initialize!");
+  analyzer_group_ = new AnalyzerGroup();
```
I strongly believe they should be unique pointer.

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -33,214 +33,285 @@
   }
 
   // Last analyzer handles remaining data
   other_analyzer_ = new OtherAnalyzer(other_as_errors);
```
unique pointer

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -33,214 +33,285 @@
-  { // lock the whole loop to ensure nothing in the analyzer group changes
-    // during it.
+  {  // lock the whole loop to ensure nothing in the analyzer group changes during it.
     boost::mutex::scoped_lock lock(mutex_);
```
should be a std::mutex

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -33,214 +33,285 @@
     boost::mutex::scoped_lock lock(mutex_);
-    for (unsigned int j = 0; j < diag_msg->status.size(); ++j)
-    {
+    for (unsigned int j = 0; j < diag_msg->status.size(); ++j) {
```
```suggestion
    for (auto j = 0u; j < diag_msg->status.size(); ++j) {
```

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -33,214 +33,285 @@
+    for (unsigned int j = 0; j < diag_msg->status.size(); ++j) {
       analyzed = false;
-      boost::shared_ptr<StatusItem> item(new StatusItem(&diag_msg->status[j]));
+      std::shared_ptr<StatusItem> item(new StatusItem(&diag_msg->status[j]));
```
make shared

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -33,214 +33,285 @@
   }
-  for (unsigned int i = 0; i < processed.size(); ++i)
-  {
+  for (unsigned int i = 0; i < processed.size(); ++i) {
```
```suggestion
  for (const auto & msg : processed) {
```

Karsten1987 commented on diagnostic_aggregator/src/aggregator_node.cpp:
```diff
@@ -34,34 +34,37 @@
+  try {
+    diagnostic_aggregator::Aggregator agg;
+
+    rclcpp::executors::SingleThreadedExecutor exec;
```
if only one node is present, there is no need for an executor.
Just call `rclcpp::spin_some(agg)`.

Karsten1987 commented on diagnostic_aggregator/src/aggregator_node.cpp:
```diff
@@ -34,34 +34,37 @@
+      agg.publishData();
+      pub_rate.sleep();
+    }
+  } catch (exception & e) {
```
```suggestion
  } catch (const exception & e) {
```

Karsten1987 commented on diagnostic_aggregator/src/analyzer_group.cpp:
```diff
@@ -33,252 +33,308 @@
+  std::string p_type = breadcrumb_.empty() ? "type" : breadcrumb_ + ".type";
+  std::string p_path = breadcrumb_.empty() ? "path" : breadcrumb_ + ".path";
+  
+  for (auto & param : parameters) {
```
```suggestion
  for (const auto & param : parameters) {
```

Karsten1987 commented on diagnostic_aggregator/src/analyzer_group.cpp:
```diff
@@ -33,252 +33,308 @@
+      pos = 10;
+    }
+    ns = param.first.substr(0, param.first.find(".", pos));
+    //std::cout << "NS: "<<ns<<std::endl<<std::endl;
```
```suggestion
```

Karsten1987 commented on diagnostic_aggregator/src/analyzer_group.cpp:
```diff
@@ -33,252 +33,308 @@
         }
+
+        analyzer = analyzer_loader_.createSharedInstance(an_type);
+      } catch (pluginlib::LibraryLoadException & e) {
```
```suggestion
      } catch (const pluginlib::LibraryLoadException & e) {
```

Karsten1987 commented on diagnostic_aggregator/src/analyzer_group.cpp:
```diff
@@ -33,252 +33,308 @@
+        RCLCPP_ERROR(get_logger(
+            "AnalyzerGroup"), "Failed to load analyzer %s, type %s. Caught exception: %s",
+          ns.c_str(), an_type.c_str(), e.what());
+        std::shared_ptr<StatusItem> item(new StatusItem(ns,
```
make shared

Karsten1987 commented on diagnostic_aggregator/src/analyzer_group.cpp:
```diff
@@ -33,252 +33,308 @@
   return init_ok;
 }
 
 AnalyzerGroup::~AnalyzerGroup()
```
virtual

Karsten1987 commented on diagnostic_aggregator/src/analyzer_group.cpp:
```diff
@@ -33,252 +33,308 @@
-  if (it != analyzers_.end())
-  {
+  RCLCPP_DEBUG(get_logger("AnalyzerGroup"), "removeAnalyzer()");
+  vector<std::shared_ptr<Analyzer>>::iterator it = find(analyzers_.begin(),
```
```suggestion
  auto it = find(analyzers_.begin(),
```

Karsten1987 commented on diagnostic_aggregator/src/analyzer_group.cpp:
```diff
@@ -33,252 +33,308 @@
+  // First check cache
+  if (matched_.count(name)) {
+    vector<bool> & mtch_vec = matched_[name];
+    for (unsigned int i = 0; i < mtch_vec.size(); ++i) {
```
consider using `std::find` for this.

Karsten1987 commented on diagnostic_aggregator/src/analyzer_group.cpp:
```diff
@@ -33,252 +33,308 @@
 
     // Do not report anything in the header values for analyzers that don't report
-    if (processed.size() == 0)
+    if (processed.size() == 0) {
```
```suggestion
    if (processed.empty()) {
```

Karsten1987 commented on diagnostic_aggregator/src/generic_analyzer.cpp:
```diff
@@ -33,237 +33,310 @@
+  int num_items_expected;
+  bool discard_stale;
+
+  for (auto & param : parameters) {
```
```suggestion
  for (const auto & param : parameters) {
```

norro replied:
Good point.

Karsten1987 commented on diagnostic_aggregator/src/generic_analyzer.cpp:
```diff
@@ -33,237 +33,310 @@
+        nice_name_.c_str(), pvalue.value_to_string().c_str());
+      for (auto regex : pvalue.as_string_array()) {
+        try {
+          boost::regex re(regex);
```
std::regex

Karsten1987 commented:
besides the nitpicks, the code as is doesn't compile on my osx machine. Mainly because it doesn't find boost, but also the `AnalyzerLoader` test doesn't compile because line 52 has not enough parameters.

Besides all that, I would like to know what's the reason for removing all the variable names in the function declarations. That's a bit tricky to debug currently.

Karsten1987 reviewed (changes requested):
Code currently doesn't compile

norro committed:
Fixes complation of analyzer group test

norro commented:
Thank you very much for the _nitpicks_, i.e. your detailed review, @Karsten1987. This is very much appreciated!
I forgot to commit the changes to the test, this should compile now as well.

Regarding the variable names in the function declarations, I think I found both versions in the existing code (with and without variable names) and wanted to have it consistent in the package.
I obviously decided for the option that makes the review harder, which wasn't my intention. Also, this is not consistent with the other diagnostic packages, so it wasn't a wise decision, I suppose.
I probably have to go over the documentation blocks in the headers anyway to check their consistence with the function declarations. I can add the names in that commit as well.

norro committed:
Removed dependency to boost

* Using std::mutex instead of boost::mutex. Using std::lock_guard
instead of boost::scoped_lock since std::scoped_lock was not introduced before C++17
* Using std::regex instead of boost::regex

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Alphabetical order of includes and dependencies

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Adopted suggestions from review by @Karsten1987

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

Karsten1987 commented on diagnostic_aggregator/test/aggregator_test.py.stashed:
```diff
@@ -0,0 +1,274 @@
+#!/usr/bin/python3
```
should that file be committed?

norro committed:
Minor improvements

* Using unique_ptrs instead of plain c pointers
* Simplifying loops

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

Karsten1987 commented:
As discussed offline, I've continued working on this branch. However, I can't easily merge this PR as it is actually target the `updater` branch instead of `ros2-devel`. Also, I've seen that there is already a `migrate_diagnostic_aggregator` branch. So I am going ahead and close this PR and push your changes to that upstream branch.
